### ISSUE TITLE
Add type hints

### ISSUE THREAD

**Issue by devRMA** [JulienPalard/Pipe#76]
Use a static type checker called [mypy](https://github.com/python/mypy). In libs, I think it's very important to have type hints, so that when users are going to use the lib, code linters can help, knowing exactly what kind of parameters to receive, and what the return type is.

The ideal solution would be to add type hints as specified by [PEP 484](https://www.python.org/dev/peps/pep-0484/), [PEP 526](https://www.python.org/dev/peps/pep-0526/), [PEP 544](https://www.python.org/dev/peps/pep-0544/), [PEP 586](https://www.python.org/dev/peps/pep-0586/), [PEP 589](https://www.python.org/dev/peps/pep-0589/), and [PEP 591](https://www.python.org/dev/peps/pep-0591/) directly on your code.

**Comment by devRMA** [comment]
I even noticed that you had already added mypy to the requirements. Just need to do static typing.

https://github.com/JulienPalard/Pipe/blob/dd179c8ff0aa28ee0524f3247e5cb1c51347cba6/requirements.txt#L3-L4

**Comment by JulienPalard** [comment]
Probably blocked by https://github.com/python/mypy/issues/11833, as I'll need something like:

```python
P = ParamSpec("P")
I = TypeVar("I")
O = TypeVar("O")

class Pipe(Generic[P, I, O]):
    def __init__(self, function: Callable[Concatenate[Iterable[I], P], Iterator[O]]) -> None:
        ...
```

as later the `__call__` arguments are `P`. (The `Iterable[I]` being passed to the function via `__ror__`.)

But yes I agree it would be very nice to be able to write something like:

```python
T = TypeVar("T")
take: Pipe[T, int, T]
```

to express than `take` takes an `int`, and yields the same type that the feeded type.

**Comment by JulienPalard** [comment]
Now blocked by mypy issue 1484 as we need a way to cleanly decribe functools.partial, to properly describe:

    def __call__(self, *args, **kwargs):
        return Pipe(
            lambda iterable, *args2, **kwargs2: self.function(
                iterable, *args, *args2, **kwargs, **kwargs2
            )
        )

**Comment by 0xMochan** [comment]
> Now blocked by mypy issue 1484 as we need a way to cleanly decribe functools.partial, to properly describe:
> 
> ```
> def __call__(self, *args, **kwargs):
>     return Pipe(
>         lambda iterable, *args2, **kwargs2: self.function(
>             iterable, *args, *args2, **kwargs, **kwargs2
>         )
>     )
> ```

I'm pretty sure this is nearly impossible without being able to represent HKTs (Higher-Kinded Types) in the Python type system.

**Comment by fabiob** [comment]
@JulienPalard and @devRMA, would you share the code you write so far in a branch or fork? I would love to take a swing at it and try to help you.

**Comment by JulienPalard** [comment]
@fabiob pushed my `typing` branch, see the `pyi` file. I know it's not much and does not works.

What we really need is a clean way to type `functools.partial`, as it's what's `pipe` does a lot.

**Comment by p-98** [comment]
Seems like https://github.com/python/mypy/issues/1484 was completed with https://github.com/python/mypy/pull/16939.

Does the implementation provide what you need?

**Comment by JulienPalard** [comment]
I don't think so, they implemented a mypy plugin to handle the single `functools.partial` function, I really mean there's an actual `if fullname == "functools.partial":` in the implementation.

Looks like there's still no way to type annotate `functools.partial` equivalents.

**Comment by JulienPalard** [comment]
Currently this works:

```python
from collections.abc import Callable
from typing import Concatenate, Generic, ParamSpec, TypeVar

P = ParamSpec('P')
T = TypeVar('T')
U = TypeVar('U')

class Pipe(Generic[U, P, T]):
    def __init__(self, function: Callable[Concatenate[U, P], T]):
        self.function = function

    def run(self, arg1: U) -> T:
        return self.function(arg1)


def simple_annotated_function(value: int) -> str:
    assert isinstance(value, int)
    return "hello"


def main() -> None:
    p = Pipe(simple_annotated_function)
    reveal_type(p)  # demo.Pipe[builtins.int, [], builtins.str]
    reveal_type(p.function) # def (builtins.int) -> builtins.str
    reveal_type(p.run) # def (arg1: builtins.int) -> builtins.str
    p.run("coucou") # Argument 1 to "run" of "Pipe" has incompatible type "str"; expected "int"


main()
```

But it feels like it's only 1/10 of the work, as we need some way to "substract" parameters from a paramspec.

**Comment by JulienPalard** [comment]
Starting a new discussion here: https://github.com/python/mypy/issues/17481 to track this.

**Comment by EGAMAGZ** [comment]
Are you still considering or are you already working on adding typing to the library?

**Comment by JulienPalard** [comment]
> Are you still considering or are you already working on adding typing to the library?

I'm stuck, see the messages in this issue, any help welcome :)

**Comment by BoggerByte** [comment]
Hi! @JulienPalard I’ve just opened a PR that introduces type hints across the project as a first beta pass: #117.

It includes support for most built-in pipes, a new `cast` pipe, and some exploratory typing patterns like overloads and generics. There are still a few edge cases and unresolved mypy quirks (e.g. in `chain_with` and `izip`), but overall the typing is functional and editor-friendly.

Would love feedback, especially on `Any` vs `object` usage, and on whether the `cast` pipe needs runtime type checking.

**Comment by vthemelis** [comment]
@JulienPalard i also just tried to typecheck this library and got stuck with not having a union of ParamSpec available.

It may be good to dual post your mypy request to pyright as I find it tends to be a bit more mature and nuanced than mypy. I'm happy to do this for you if you prefer.

**Comment by JulienPalard** [comment]
> It may be good to dual post your mypy request to pyright as I find it tends to be a bit more mature and nuanced than mypy. I'm happy to do this for you if you prefer.

Good idea! I'm no pyright user, you can post it if you want.
